src/compat.jl

```diff
@@ -0,0 +1,373 @@
+    id = newnode!(graph)
+    sethead!(graph, id, k)
+    setattr!(graph, id, source=src)
+    setflags!(graph, id, flags)
+    return id
+end
+
+"""
+An Expr -> SyntaxTree transformation that should preserve semantics, but will
+have low-quality provenance info (namely, each tree node will be associated with
+the last seen LineNumberNode in the pre-order expr traversal).
+
+Last-resort option so that, for example, we can lower the output of old
+Expr-producing macros.  Always prefer re-parsing source text over using this.
+
+Does not support lowered exprs.
+"""
+function expr_to_syntaxtree(@nospecialize(e), lnn::Union{LineNumberNode, Nothing}=nothing)
+    graph = SyntaxGraph()
+    ensure_attributes!(graph,
+                       kind=Kind, syntax_flags=UInt16,
+                       source=SourceAttrType, var_id=Int, value=Any,
+                       name_val=String, is_toplevel_thunk=Bool)
+    toplevel_src = if isnothing(lnn)
+        # Provenance sinkhole for all nodes until we hit a linenode
+        dummy_src = SourceRef(
+            SourceFile("No source for expression $e"),
+            1, JuliaSyntax.GreenNode(K"None", 0))
+        _insert_tree_node(graph, K"None", dummy_src)
+    else
+        lnn
+    end
+    st_id, _ = _insert_convert_expr(e, graph, toplevel_src)
+    out = SyntaxTree(graph, st_id)
+    return out
+end
+
+function _expr_replace!(@nospecialize(e), replace_pred::Function, replacer!::Function,
+                        recurse_pred=(@nospecialize e)->true)
+    if replace_pred(e)
+        replacer!(e)
+    end
+    if e isa Expr && recurse_pred(e)
+        for a in e.args
+            _expr_replace!(a, replace_pred, replacer!, recurse_pred)
+        end
+    end
+end
+
+function _to_iterspec(exs::Vector)
+    if length(exs) === 1 && exs[1].head === :filter
+        @assert length(exs[1].args) >= 2
+        return Expr(:filter, _to_iterspec(exs[1].args[2:end]), exs[1].args[1])
+    end
+    outex = Expr(:iteration)
+    for e in exs
+        if e.head === :block
+            for iter in e.args
+                push!(outex.args, Expr(:in, iter.args...))
+            end
+        elseif e.head === :(=)
+            push!(outex.args, Expr(:in, e.args...))
+        else
+            @assert false "unknown iterspec in $e"
+        end
+    end
+    return outex
+end
+
+"""
+Return `e.args`, but with any parameters in SyntaxTree (flattened, source) order.
+Parameters are expected to be as `e.args[pos]`.
+
+e.g. orderings of (a,b,c;d;e;f):
+  Expr:       (tuple (parameters (parameters (parameters f) e) d) a b c)
+  SyntaxTree: (tuple a b c (parameters d) (parameters e) (parameters f))
+"""
+function collect_expr_parameters(e::Expr, pos::Int)
+    params = expr_parameters(e, pos)
+    isnothing(params) && return e.args
+    args = Any[e.args[1:pos-1]..., e.args[pos+1:end]...]
+    return _flatten_params(params, args)
+end
+function _flatten_params(p::Expr, out::Vector{Any})
```

Comment:
```suggestion
function _flatten_params!(out::Vector{Any}, p::Expr)
```
just to follow mutation convention